[clang-cl] [Sema] Support MSVC non-const lvalue to user-defined temporary reference

MaxEW707 opened the pull request:
MSDN docs for reference: https://learn.microsoft.com/en-us/cpp/build/reference/zc-referencebinding-enforce-reference-binding-rules?view=msvc-170
The warning referenced in that MSDN article: https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-4-c4239?view=msvc-170

MSVC has an extension, that has been enabled by default until very recently when MSVC started their C++ conformance trend, that allows binding a user-defined type temporary to a non-const lvalue.

When trying to port over various large internal windows only tools this became hard as they rely heavily on this extension.
While I am not a fan of this extension changing our code to remove uses of it is quite difficult to do because of how the code is structured.
The end goal is to no longer rely on this extension but at this moment in time it is less bug prone to have clang support this extension and focus on removal sometime in the future after porting to clang.

From the MSDN reference above I did not implement the warning at level 4, C4239, since everyone disables said warning who requires this extension so I felt it wasn't worth the effort. Implementing the MSVC extension is the main requirement.

I also enable this new option with `/permissive` since `/permissive` enables this MSVC extension on MSVC.

MaxEW707 committed:
Support MSVC lvalue to temporary reference binding

MaxEW707 committed:
update description of lang opt

rnk commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -5138,7 +5148,11 @@ TryReferenceInit(Sema &S, Expr *Init, QualType DeclType,
   //        non-volatile const type (i.e., cv1 shall be const), or the reference
   //        shall be an rvalue reference.
-  if (!isRValRef && (!T1.isConstQualified() || T1.isVolatileQualified())) {
+  const bool CanBindLValueRef =
```
This boolean isn't needed if `!isRValRef`. Can you structure it as:
```
  if (!isRValRef) {
    bool CanBindLValueRef = ...
    if (!CanBindLValueRef) {
    }
  }
```
... or whatever is most readable, I just want to reduce the scope of the variable and avoid extra LangOpt checks.

rnk reviewed (approved):
I think this looks good, but I would appreciate another reviewer looking at the patch.

mizvekov commented on clang/lib/Sema/SemaInit.cpp:
```diff
@@ -5213,13 +5213,17 @@ static void TryReferenceInitializationCore(Sema &S,
-              : InitializationSequence::FK_ReferenceInitDropsQualifiers);
-    else {
+    else if (!InitCategory.isLValue()) {
+      if (T1Quals.isAddressSpaceSupersetOf(T2Quals)) {
+        if (!S.getLangOpts().MSVCReferenceBinding ||
+            !S.AllowMSLValueReferenceBinding(T1Quals, T1))
+          Sequence.SetFailed(InitializationSequence::
+                                 FK_NonConstLValueReferenceBindingToTemporary);
```
So we were in a branch where this condition holds: `!(T1Quals.hasConst() && !T1Quals.hasVolatile() && T1Quals.isAddressSpaceSupersetOf(T2Quals))`.

Now we enter a branch where `T1Quals.isAddressSpaceSupersetOf(T2Quals)` holds.

Together these two premises imply that `T1Quals.hasConst() && !T1Quals.hasVolatile()` doesn't hold, which is the same as to say `!T1Quals.hasConst() || T1Quals.hasVolatile()` doesn't hold.

This implies that the `AllowMSLValueReferenceBinding` call can only be true iff `Quals.hasVolatile() || QT->isBuiltinType() || QT->isArrayType()` is false, making the `Quals.hasConst()` check in AllowMSLValueReferenceBinding superfluous for this call site.

On the other call site, `T1.isConstQualified() && !T1.isVolatileQualified()` already implies that `AllowMSLValueReferenceBinding` call will return true.

So if I haven't made any mistakes so far (please double check!) you could pretty much perform this rewrite in `AllowMSLValueReferenceBinding`:
```C++
bool Sema::AllowMSLValueReferenceBinding(Qualifiers Quals, QualType QT) {
   return getLangOpts().MSVCReferenceBinding && !(Quals.hasVolatile() || QT->isBuiltinType() || QT->isArrayType());
 }
 ```

Which now justifies this function being a non-static member in Sema, since before it did not need `this` at all.

With this change, it seems the changes needed on the second call site will look much simpler as well, just a one liner:
```C++
if (!isRValRef && (!T1.isConstQualified() || T1.isVolatileQualified() || S.AllowMSLValueReferenceBinding(T1.getQualifiers(), T1))) {
```
But again, please double check!

MaxEW707 replied:
You are correct. I reorder the logic as you have described :).

mizvekov commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -10158,6 +10158,8 @@ class Sema final : public SemaBase {
   CompareReferenceRelationship(SourceLocation Loc, QualType T1, QualType T2,
                                ReferenceConversions *Conv = nullptr);
 
+  bool AllowMSLValueReferenceBinding(Qualifiers Quals, QualType QT);
```
In case you don't implement my other suggestion, this can be static.
```suggestion
  static bool AllowMSLValueReferenceBinding(Qualifiers Quals, QualType QT);
```

MaxEW707 committed:
PR Feedback on SemaOverload.cpp if check variable scoping

MaxEW707 committed:
Reorder logic around `AllowMSLValueReferenceBinding`

shafik reviewed (commented):
I would like @AaronBallman to review this before committing.

AaronBallman reviewed (changes requested):
Thank you for this! I think there's some work left to be done though.

1) Please add an extension warning (and tests for it) so users know where the extension is being used even if they opt into it explicitly. This is important for `-pedantic` mode.
2) This should be controllable via `-fms-extensions`/`-fno-ms-extensions`; having its own dialect flag is a bit novel but I'm not strongly opposed. CC @MaskRay @jansvoboda11 for driver/options opinions
3) Please document the extension in `clang/docs/LanguageExtensions.rst`

MaxEW707 commented:
> This should be controllable via -fms-extensions/-fno-ms-extensions; having its own dialect flag is a bit novel but I'm not strongly opposed. CC @MaskRay @jansvoboda11 for driver/options opinions

For this feedback I intentionally didn't do this because this ms extension isn't one intended to be always enabled.
-fms-extensions enables extensions that are always valid to use with msvc such as the `__try` and `__finally` keywords for SEH.

MSVC allows controlling this option independently here, [/Zc:ReferenceBinding](https://learn.microsoft.com/en-us/cpp/build/reference/zc-referencebinding-enforce-reference-binding-rules?view=msvc-170).
MSVC `/permissive-` disables this reference binding extension, however `__try` and friends are still available as they are vital for operating within a Win32 environment. Newer C++ modes like C++20 also implicitly disable this reference binding extension in MSVC as they move towards proper C++ conformance by default.

This is a one off extension similar to -fms-volatile or -fms-define-stdc. I would like to keep it a separate option that can be individually controlled instead of grouping it with -fms-extensions.

AaronBallman commented:
> > This should be controllable via -fms-extensions/-fno-ms-extensions; having its own dialect flag is a bit novel but I'm not strongly opposed. CC @MaskRay @jansvoboda11 for driver/options opinions
> 
> For this feedback I intentionally didn't do this because this ms extension isn't one intended to be always enabled. -fms-extensions enables extensions that are always valid to use with msvc such as the `__try` and `__finally` keywords for SEH.

That's kind of the crux of this -- `-fms-extensions` enables Microsoft extensions, and this is a Microsoft extension. Our documentation is pretty quiet about what it means to "enable Microsoft extensions": https://clang.llvm.org/docs/UsersManual.html#microsoft-extensions

> MSVC allows controlling this option independently here, [/Zc:ReferenceBinding](https://learn.microsoft.com/en-us/cpp/build/reference/zc-referencebinding-enforce-reference-binding-rules?view=msvc-170). MSVC `/permissive-` disables this reference binding extension, however `__try` and friends are still available as they are vital for operating within a Win32 environment. Newer C++ modes like C++20 also implicitly disable this reference binding extension in MSVC as they move towards proper C++ conformance by default.
> 
> This is a one off extension similar to -fms-volatile or -fms-define-stdc. I would like to keep it a separate option that can be individually controlled instead of grouping it with -fms-extensions.

Oh, I think it should still be individually controlled via its own flag. The scenario I am thinking about is when the user passed `-fms-extensions` and no other individual flags. I think that mode should enable all of the Microsoft extensions. Then users can opt out of whatever extensions they don't want to enable. But I think it's confusing if `-fms-extensions` still requires the user to explicitly opt in to other Microsoft-specific extensions. CC @zmodem @rnk for additional opinions.

MaxEW707 committed:
PR Feedback on ext warning

MaxEW707 committed:
quoting

MaxEW707 committed:
Fix doc build

MaxEW707 committed:
Default ms reference binding to on unless C++20 similar to delayed-template-parsing

MaxEW707 commented:
> The scenario I am thinking about is when the user passed -fms-extensions and no other individual flags. I think that mode should enable all of the Microsoft extensions. Then users can opt out of whatever extensions they don't want to enable. But I think it's confusing if -fms-extensions still requires the user to explicitly opt in to other Microsoft-specific extensions.

That is a fair point. I did some further digging.

Starting with C++20 MSVC starting enforcing various C++ conformance options by default. For example `/permissive-` is the default now meaning `/Zc:referenceBinding` is set. MSVC does not allow its custom reference binding rules by default anymore.
Godbolt for reference: https://godbolt.org/z/6hTeGeb6x.

We already do similar behaviour in the driver for the `delayed-template-parsing` extension.
`delayed-template-parsing`, called `/Zc:twoPhase-` in MSVC driver parlance, is enabled by default if we are windows msvc and not C++20 since MSVC does two-phase lookup by default in C++20 onwards.

In this PR I now enable `ms-reference-binding` by default if we are windows msvc and not C++20 as well to imitate MSVC behaviour.

`ms-reference-binding` is still an individual controllable option and is also properly set if you use the `/permissive` or `/permissive-` larger group option in the cl driver mode.

I wrote up a small gist here about what `-fms-extensions` covers, https://gist.github.com/MaxEW707/3fbb3a223710b6aa9269f7ffbdccbfc9.
Most of it are things that you always want if you are targeting msvc which include the implicit `_GUID` struct, calling convention identifiers like `__stdcall`, SEH identifiers like `__finally`, Microsoft pragma support, builtin macros like `_WCHAR_T_DEFINED`, imitating MSVC PP token pasting behaviour, handling ^Z EOF in the lexer, etc.
It also includes Microsoft extensions that do not have a conformance mode toggle currently in MSVC such as allowing a static declaration to follow an extern declaration, allowing inline function definitions on pure virtual methods, allowing type definitions inside an anonymous union/struct, etc. Most of these extensions in this category so far are fairly benign and don't alter the language semantics much compared to two-phase lookup or reference binding for example.

Let me know if this is a satisfactory compromise :). I think it models what seems to be the precedent especially around two-phase lookup.

MaxEW707 committed:
remove old comment

MaxEW707 committed:
Fix overload resolution and only consider msvc binding if we are an rvalue

MaxEW707 committed:
fix comment

MaxEW707 commented on clang/lib/Sema/SemaOverload.cpp:
```diff
@@ -4446,6 +4446,24 @@ CompareStandardConversionSequences(Sema &S, SourceLocation Loc,
+        // When binding a user-defined type temporary to an lvalue MSVC will
+        // pick `const T&` over `T&` when binding an unqualified `T&&`.
+        // Therefore we want to pick the more qualified const overload in this
+        // specific case.
```
https://godbolt.org/z/xaPE6rMc1 for reference on this overload resolution case.

MaxEW707 committed:
Fix CXX method functions

AaronBallman commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -5821,3 +5821,33 @@ specify the starting offset to begin embedding from. The resources is treated
+MSVC Extensions
+===============
+
+Clang supports a number of extensions inorder to imitate MSVC.
```
```suggestion
Clang supports a number of extensions in order to imitate MSVC.
```

AaronBallman commented on clang/test/SemaCXX/ms-reference-binding.cpp:
```diff
@@ -0,0 +1,151 @@
+  NS::fIntConstArray({ 1 });
+}
+
+#endif
```
Some other test coverage to consider:

* Binding to a type that's a `typedef` to a class type
* Dependent types
* Function types
* Attributed types (such as a calling convention on a function type)

RIscRIpt replied:
* Default arguments https://godbolt.org/z/YEqsE4GzG

AaronBallman reviewed (commented):
Thanks, I think this is a good compromise! I did have some additional test coverage, and I'd appreciate additional eyes from @zmodem and @MaskRay, especially regarding the compiler options.

zmodem commented on clang/docs/LanguageExtensions.rst:
```diff
@@ -5821,3 +5821,33 @@ specify the starting offset to begin embedding from. The resources is treated
+As of Visual Studio 2017, ``/permissive-`` is the default for projects meaning C++ conformance is enforced when
+building with MSVC in Visual Studio.
+
+This MSVC extension can be enabled with ``-fms-reference-binding`` with the clang or cl driver.
+This MSVC extension can be enabled with ``/Zc:referenceBinding-`` with the cl driver.
```
Let's s/cl/clang-cl/g. (I think we always refer to it as clang-cl in docs.)

zmodem commented on clang/include/clang/Driver/Options.td:
```diff
@@ -3034,6 +3034,12 @@ def fms_extensions : Flag<["-"], "fms-extensions">, Group<f_Group>,
   MarshallingInfoFlag<LangOpts<"MicrosoftExt">>, ImpliedByAnyOf<[fms_compatibility.KeyPath]>;
+def fms_reference_binding : Flag<["-"], "fms-reference-binding">, Group<f_Group>,
+  Visibility<[ClangOption, CC1Option, CLOption]>,
+  HelpText<"Accept expressions that bind a non-const lvalue reference to a user-defined type temporary as supported by the Microsoft Compiler">,
```
s/Compiler/compiler/

I know the file is not consistent, but historically it's been trying to stay within 80 columns, so consider breaking up the HelpText (also below). See e.g. fms_compatibility_version for an example.

zmodem commented on clang/include/clang/Driver/Options.td:
```diff
@@ -8492,6 +8498,12 @@ def _SLASH_Zc_wchar_t : CLFlag<"Zc:wchar_t">,
 def _SLASH_Zc_wchar_t_ : CLFlag<"Zc:wchar_t-">,
   HelpText<"Disable C++ builtin type wchar_t">;
+def _SLASH_Zc_referenceBinding : CLFlag<"Zc:referenceBinding">,
+  HelpText<"Do not accept expressions that bind a non-const lvalue reference to a user-defined type temporary">,
```
"(default)" at the end of the help text.

zmodem commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -7272,6 +7272,11 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
 
+  if (Args.hasFlag(options::OPT_fms_reference_binding,
+                   options::OPT_fno_ms_reference_binding,
+                   IsWindowsMSVC && !HaveCxx20))
```
Could we do something better for the non-windows or c++20 cases?

We could emit a warning saying the flag is unsupported, or we could put those conditions first so that we don't call hasFlag, which would make clang warn about the unused option.

MaxEW707 replied:
I am not quite following. I modelled this flag similar to how we do other ms flags such as `-fms-extensions` and `-fdelayed-template-parsing`. That is it is enabled by cl driver options or enabled by default for MSVC targets.
It is still valid for you to explicitly do `-fms-extensions` on non-msvc targets if you want those behaviours for reasons.

When porting some old MSVC only games to certain console manufacturers it was common to use `-fms-extensions` and friends to get it building on their vendor clang :).

zmodem commented on clang/lib/Driver/ToolChains/MSVC.cpp:
```diff
@@ -950,12 +950,14 @@ static void TranslatePermissive(Arg *A, llvm::opt::DerivedArgList &DAL,
                                 const OptTable &Opts) {
   DAL.AddFlagArg(A, Opts.getOption(options::OPT__SLASH_Zc_twoPhase_));
   DAL.AddFlagArg(A, Opts.getOption(options::OPT_fno_operator_names));
+  DAL.AddFlagArg(A, Opts.getOption(options::OPT_fms_reference_binding));
```
Do we need to check for c++20 like we do above?

zmodem reviewed (commented):
Just some nits from my side. (I only looked at the docs and driver code)

RogerSanders commented:
I'd like to see this change make it into the repo, as I'm in a similar situation with a 6+ million line codebase. What's the next steps to advance this?

MaxEW707 commented:
> I'd like to see this change make it into the repo, as I'm in a similar situation with a 6+ million line codebase. What's the next steps to advance this?

Hey sorry. I've been busy with other work prios and an msvc mangling bug. I still intend to get back to this as I also need it.
I plan to get back to it in about 1.5 weeks to address all the feedback.

I also want to change it to be opt-in for clang-cl for now since it is the first impl instead of being opt-out like in msvc.
I don't want to accidentally break users who are using clang-cl without `/permissive-` especially since it changes overload resolution behaviour.

MaxEW707 committed:
Disable reference binding by default and fix options comments

MaxEW707 committed:
Fix inorder spelling

MaxEW707 committed:
cl -> clang-cl

MaskRay commented on clang/lib/Driver/ToolChains/Clang.cpp:
```diff
@@ -7381,6 +7381,11 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
     CmdArgs.push_back("-fdelayed-template-parsing");
   }
 
+  if (Args.hasFlag(options::OPT_fms_reference_binding,
```
Simplify with `addOptInFlag`

MaxEW707 committed:
Add unit tests for default arguments

MaxEW707 committed:
Add another default arg unit test

MaxEW707 committed:
Convert to addOptInFlag

MaxEW707 committed:
Typedef lvalue ref tests

MaxEW707 committed:
Add ref related derived to base cv binding test

MaxEW707 commented on clang/test/CodeGenCXX/ms-reference-binding.cpp:
```diff
@@ -0,0 +1,87 @@
+void fAPickRef(A&) {}
+void fAPickRef(const volatile A&) {}
+
+// NOTE: MSVC incorrectly picks the `const volatile A&` overload with the mangled name
```
https://godbolt.org/z/GvWY4n7rK

https://developercommunity.visualstudio.com/t/_MSC_VER--1940-MSVC-allows-binding-a/10811594

As noted in the comment this appears to be a bug in MSVC here as it also occurs during conforming reference binding. For clang we will continue to pick the `A&` overload when ms non-conforming reference binding is in use.

MaxEW707 committed:
Add conversion and temporary materialization unit tests; enforece variable non-const lvalue reference init doesn't consider ref incompatible relationships

MaxEW707 committed:
Add unit test for reference to pointers

MaxEW707 committed:
Fix expected-note in unit test

MaxEW707 committed:
Add member pointer tests; Add vcall tests

MaxEW707 committed:
Simplify reference binding check to just `RecordType`

MaxEW707 committed:
Add template unit tests; Add unit tests for vector types

MaxEW707 committed:
explicit template instantiation test

MaxEW707 commented:
@shafik 
@AaronBallman 
@zmodem 

I believe I addressed all the feedback. Let me know if I missed anything.
Feel free to ping anyone else I may have missed above :).
